File: scripts/ops/execute-push-verify.ts
#!/usr/bin/env tsx
/**
 * Execute push and verify commands directly
 */

import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

const SESSION_PATH = path.join(process.cwd(), 'twitter_session.json');
const RAILWAY_SERVICE = 'serene-cat';

async function main() {
  console.log('🚀 Pushing Session to Railway');
  console.log('═══════════════════════════════════════════════════════════\n');
  
  // Verify session file exists
  if (!fs.existsSync(SESSION_PATH)) {
    console.error(`❌ Session file not found: ${SESSION_PATH}`);
    process.exit(1);
  }
  
  const sessionData = fs.readFileSync(SESSION_PATH, 'utf8');
  console.log(`📁 Session file: ${SESSION_PATH} (${sessionData.length} bytes)`);
  
  // Check for auth cookies
  try {
    const session = JSON.parse(sessionData);
    const cookies = session.cookies || [];
    const authToken = cookies.find((c: any) => c.name === 'auth_token');
    const ct0 = cookies.find((c: any) => c.name === 'ct0');
    
    console.log(`\n🍪 Cookie Check:`);
    console.log(`   auth_token: ${authToken ? '✅ YES' : '❌ NO'}`);
    console.log(`   ct0: ${ct0 ? '✅ YES' : '❌ NO'}\n`);
    
    if (!authToken || !ct0) {
      console.error('❌ Missing required auth cookies!');
      process.exit(1);
    }
  } catch (err) {
    console.error('❌ Invalid JSON in session file');
    process.exit(1);
  }
  
  // Base64 encode
  const sessionB64 = Buffer.from(sessionData).toString('base64');
  console.log(`📦 Encoded to base64: ${sessionB64.length} chars\n`);
  
  // Push to Railway
  console.log(`🚀 Updating Railway service: ${RAILWAY_SERVICE}`);
  try {
    execSync(`railway variables --service ${RAILWAY_SERVICE} --set "TWITTER_SESSION_B64=${sessionB64}"`, {
      stdio: 'inherit',
      encoding: 'utf8',
    });
    console.log('✅ Railway variable updated\n');
  } catch (error: any) {
    console.error(`❌ Failed to update Railway: ${error.message}`);
    process.exit(1);
  }
  
  // Verify auth
  console.log('🔍 Verifying auth on Railway...');
  console.log('═══════════════════════════════════════════════════════════\n');
  try {
    const output = execSync(
      `railway run --service ${RAILWAY_SERVICE} pnpm exec tsx scripts/ops/run-harvester-single-cycle.ts`,
      {
        encoding: 'utf8',
        maxBuffer: 10 * 1024 * 1024, // 10MB
        timeout: 120000, // 2 minutes
      }
    );
    
    const authLines = output.split('\n').filter((line: string) => line.includes('[HARVESTER_AUTH]'));
    
    if (authLines.length > 0) {
      console.log('📋 [HARVESTER_AUTH] lines:');
      authLines.forEach((line: string) => console.log(`   ${line}`));
      
      const loggedIn = authLines.some((line: string) => line.includes('logged_in=true'));
      if (loggedIn) {
        console.log('\n✅ PASS: logged_in=true');
        process.exit(0);
      } else {
        console.log('\n❌ FAIL: logged_in=false');
        process.exit(1);
      }
    } else {
      console.log('⚠️  No [HARVESTER_AUTH] lines found in output');
      console.log('\nFull output (last 50 lines):');
      const lines = output.split('\n');
      lines.slice(-50).forEach((line: string) => console.log(line));
      process.exit(1);
    }
  } catch (error: any) {
    console.error(`❌ Verification failed: ${error.message}`);
    if (error.stdout) {
      const authLines = error.stdout.split('\n').filter((line: string) => line.includes('[HARVESTER_AUTH]'));
      if (authLines.length > 0) {
        console.log('\n[HARVESTER_AUTH] lines from stdout:');
        authLines.forEach((line: string) => console.log(`   ${line}`));
      }
    }
    process.exit(1);
  }
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: scripts/ops/execute-push-verify.py
#!/usr/bin/env python3
import subprocess
import json
import base64
import sys
import os

os.chdir('/Users/jonahtenner/Desktop/xBOT')

SESSION_PATH = './twitter_session.json'
RAILWAY_SERVICE = 'serene-cat'

print('🚀 Pushing Session to Railway')
print('═══════════════════════════════════════════════════════════\n')

# Verify session file
if not os.path.exists(SESSION_PATH):
    print(f'❌ Session file not found: {SESSION_PATH}')
    sys.exit(1)

with open(SESSION_PATH, 'r') as f:
    session_data = f.read()

print(f'📁 Session file: {SESSION_PATH} ({len(session_data)} bytes)')

# Check for auth cookies
try:
    session = json.loads(session_data)
    cookies = session.get('cookies', [])
    auth_token = next((c for c in cookies if c.get('name') == 'auth_token'), None)
    ct0 = next((c for c in cookies if c.get('name') == 'ct0'), None)
    
    print(f'\n🍪 Cookie Check:')
    print(f'   auth_token: {"✅ YES" if auth_token else "❌ NO"}')
    print(f'   ct0: {"✅ YES" if ct0 else "❌ NO"}\n')
    
    if not auth_token or not ct0:
        print('❌ Missing required auth cookies!')
        sys.exit(1)
except json.JSONDecodeError:
    print('❌ Invalid JSON in session file')
    sys.exit(1)

# Base64 encode
session_b64 = base64.b64encode(session_data.encode()).decode()
print(f'📦 Encoded to base64: {len(session_b64)} chars\n')

# Push to Railway
print(f'🚀 Updating Railway service: {RAILWAY_SERVICE}')
try:
    cmd = ['railway', 'variables', '--service', RAILWAY_SERVICE, '--set', f'TWITTER_SESSION_B64={session_b64}']
    subprocess.run(cmd, check=True)
    print('✅ Railway variable updated\n')
except subprocess.CalledProcessError as e:
    print(f'❌ Failed to update Railway: {e}')
    sys.exit(1)

# Verify auth
print('🔍 Verifying auth on Railway...')
print('═══════════════════════════════════════════════════════════\n')
try:
    cmd = ['railway', 'run', '--service', RAILWAY_SERVICE, 'pnpm', 'exec', 'tsx', 'scripts/ops/run-harvester-single-cycle.ts']
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
    
    output = result.stdout + result.stderr
    auth_lines = [line for line in output.split('\n') if '[HARVESTER_AUTH]' in line]
    
    if auth_lines:
        print('📋 [HARVESTER_AUTH] lines:')
        for line in auth_lines:
            print(f'   {line}')
        
        logged_in = any('logged_in=true' in line for line in auth_lines)
        if logged_in:
            print('\n✅ PASS: logged_in=true')
            sys.exit(0)
        else:
            print('\n❌ FAIL: logged_in=false')
            sys.exit(1)
    else:
        print('⚠️  No [HARVESTER_AUTH] lines found in output')
        print('\nFull output (last 50 lines):')
        lines = output.split('\n')
        for line in lines[-50:]:
            print(line)
        sys.exit(1)
except subprocess.TimeoutExpired:
    print('❌ Verification timed out')
    sys.exit(1)
except Exception as e:
    print(f'❌ Verification failed: {e}')
    sys.exit(1)
